tests/app/main/views/test_communications.py

```diff
@@ -39,88 +52,198 @@ def test_get_page_should_only_be_accessible_to_specific_user_roles(self, role, e
-        assert file_upload_messages[0].text_content().strip() == "Last modified Saturday 3 March 2018 at 3:03am GMT"
-        assert file_upload_messages[1].text_content().strip() == "Last modified Wednesday 25 July 2018 at 3:02am BST"
+        assert tuple(
+            tuple(
+                (
+                    # date text
+                    row.xpath("normalize-space(string(./td[1]))"),
+                    # download href
+                    row.xpath("./td[2]//a/@href")[0],
+                    # download text
+                    row.xpath("normalize-space(string(./td[2]//a))"),
+                    # delete href
+                    row.xpath("./td[3]//a[normalize-space(string())='Delete']/@href")[0],
+                ) for row in table.xpath(".//tr[@class='summary-item-row']")
+            ) for table in doc.xpath("//table[./caption[normalize-space(text())='Communications present']]")
+        ) == (
+            (
+                (
+                    'Monday 1 January 2018',
+                    '/admin/communications/g-things-23/files/communication/communication-foo.pdf',
+                    'document: clarification-foo.pdf',
+                    '/admin/communications/g-things-23/delete/communication/communication-foo.pdf',
+                ),
+                (
+                    'Friday 2 February 2018',
+                    '/admin/communications/g-things-23/files/communication/communication-bar.pdf',
+                    'document: clarification-bar.pdf',
+                    '/admin/communications/g-things-23/delete/communication/communication-bar.pdf',
+                ),
+                (
+                    'Saturday 3 March 2018',
+                    '/admin/communications/g-things-23/files/communication/communication-baz.pdf',
+                    'document: clarification-baz.pdf',
+                    '/admin/communications/g-things-23/delete/communication/communication-baz.pdf',
+                ),
+            ),
+        ) + ((
+            (
+                (
+                    'Wednesday 4 April 2018',
+                    '/admin/communications/g-things-23/files/clarification/clarification-qux.pdf',
+                    'document: clarification-qux.pdf',
+                    '/admin/communications/g-things-23/delete/clarification/clarification-qux.pdf',
+                ),
+                (
+                    'Wednesday 25 July 2018',
+                    '/admin/communications/g-things-23/files/clarification/clarification-dax.pdf',
+                    'document: clarification-dax.pdf',
+                    '/admin/communications/g-things-23/delete/clarification/clarification-dax.pdf',
+                ),
+            ),
+        ) if framework_status == "open" else ())
+
+        # find the form that has our submit button
+        form = doc.xpath(
+            "//form[@method='post'][@enctype='multipart/form-data'][.//input[@type='submit'][@value='Upload files']]"
```

Comment:
Maybe we should just put an id on the form? :laughing:

Reply:
But that doesn't test anything. This tests that the form is set up correctly, right method, right enctype, that there's an appropriate button...

Reply:
Ah I see, carry on then